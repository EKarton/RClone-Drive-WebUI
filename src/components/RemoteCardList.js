import RemoteCard from './RemoteCard';
import './RemoteCardList.scss';

export default function RemoteCardList({ remotes, onClick }) {
  const handleClick = (remote) => () => {
    onClick(remote);
  };

  return (
    <div className="remote-card-list">
      {remotes.sort().map((remote) => (
        <RemoteCard
          key={remote}
          className="remote-card-list__item"
          remote={remote}
          onClick={handleClick(remote)}
        />
      ))}
    </div>
  );
}
