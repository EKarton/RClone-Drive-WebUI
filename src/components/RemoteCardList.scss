.remote-card-list {
  display: flex;
  flex-wrap: wrap;
  gap: 1rem;

  &__item {
    flex-grow: 1;
  }
}
