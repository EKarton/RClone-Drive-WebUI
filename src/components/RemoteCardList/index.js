import RemoteCard from 'components/RemoteCard';
import './index.scss';

export default function RemoteCardList({ remotes, onClick }) {
  const handleClick = (remote) => () => {
    onClick(remote);
  };

  return (
    <div className="remote-card-list">
      {remotes.sort().map((remote) => (
        <RemoteCard
          key={remote}
          className="remote-card-list__item"
          remote={remote}
          onClick={handleClick(remote)}
        />
      ))}
    </div>
  );
}
