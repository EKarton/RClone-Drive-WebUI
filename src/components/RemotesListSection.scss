.remotes-list-section {
  &__header-text {
    padding-bottom: 1rem;
    font-weight: bold;
  }
}
