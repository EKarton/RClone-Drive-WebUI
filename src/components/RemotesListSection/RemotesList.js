import InfoCard from 'components/RemoteCards/InfoCard';
import RemoteCardsList from 'components/RemoteCardsList';

export default function RemotesList({ remotes, onRemoteCardClicked }) {
  const handleClick = (remote) => () => {
    onRemoteCardClicked(remote);
  };

  return (
    <RemoteCardsList>
      {remotes.sort().map((remote) => (
        <InfoCard key={remote} remote={remote} onClick={handleClick(remote)} />
      ))}
    </RemoteCardsList>
  );
}
