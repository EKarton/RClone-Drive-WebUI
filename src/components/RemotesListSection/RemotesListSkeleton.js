import SkeletonCard from 'components/RemoteCards/SkeletonCard';
import RemoteCardsList from 'components/RemoteCardsList';

export default function RemotesListSkeleton() {
  return (
    <RemoteCardsList>
      {Array.from({ length: 4 }, (_, i) => (
        <SkeletonCard key={i} />
      ))}
    </RemoteCardsList>
  );
}
