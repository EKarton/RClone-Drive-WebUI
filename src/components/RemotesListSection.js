import Typography from '@mui/material/Typography';
import RemoteCardList from 'components/RemoteCardList';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import { StatusTypes } from 'utils/constants';
import './RemotesListSection.scss';

/**
 * Represents the list of remotes with header
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes, error } = useFetchRemotes();

  if (status === StatusTypes.LOADING) {
    return null;
  }

  if (status === StatusTypes.ERROR) {
    throw error;
  }

  const handleRemoteCardClicked = (remote) => {
    onRemoteCardClicked(remote);
  };

  return (
    <>
      <Typography
        variant="h7"
        component="div"
        color="text.primary"
        className="remotes-list-section__header-text"
      >
        Your Remotes
      </Typography>
      <RemoteCardList remotes={remotes} onClick={handleRemoteCardClicked} />
    </>
  );
}
