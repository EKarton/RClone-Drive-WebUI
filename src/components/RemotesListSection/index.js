import Typography from '@mui/material/Typography';
import PropTypes from 'prop-types';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import { StatusTypes } from 'utils/constants';
import RemotesList from './RemotesList';
import RemotesListSkeleton from './RemotesListSkeleton';

/**
 * Represents the list of remotes with header
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes, error } = useFetchRemotes();

  if (status === StatusTypes.ERROR) {
    throw error;
  }

  const renderList = () => {
    if (status === StatusTypes.LOADING) {
      return <RemotesListSkeleton />;
    }

    return <RemotesList remotes={remotes} onRemoteCardClicked={onRemoteCardClicked} />;
  };

  return (
    <>
      <Typography variant="h7" component="div" color="text.primary">
        <strong>Your Remotes</strong>
      </Typography>
      {renderList()}
    </>
  );
}

RemotesListSection.propTypes = {
  onRemoteCardClicked: PropTypes.func,
};
