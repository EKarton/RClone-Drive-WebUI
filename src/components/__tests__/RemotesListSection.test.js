import RemoteCardList from 'components/RemoteCardList';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import { StatusTypes } from 'utils/constants';
import { mockRemotes } from 'test-utils/mock-responses';
import { customRender, waitFor } from 'test-utils/react';
import RemotesListSection from '../RemotesListSection';

jest.mock('hooks/fetch-data/useFetchRemotes');
jest.mock('components/RemoteCardList');

describe('RemotesListSection', () => {
  beforeEach(() => {
    useFetchRemotes.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockRemotes.remotes,
    });

    RemoteCardList.mockImplementation(({ remotes }) => {
      return <div>{remotes.join(',')}</div>;
    });
  });

  it('should match snapshot when api call was successful', async () => {
    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it('should render app boundary when api call fails', async () => {
    useFetchRemotes.mockReturnValue({
      status: StatusTypes.ERROR,
      error: new Error('Error!'),
    });

    expect(renderComponent).toThrowError();
  });

  it('should match snapshot when api call is loading', async () => {
    useFetchRemotes.mockReturnValue({
      status: StatusTypes.LOADING,
    });

    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it('should go to the files page when user clicks on a remote', async () => {
    // Mock the timer to prevent calling onClick immediately
    jest.useFakeTimers();

    RemoteCardList.mockImplementation(({ remotes, onClick }) => {
      setTimeout(() => onClick('googledrive'), 10000);
      return <div>{remotes.join(',')}</div>;
    });

    const component = renderComponent();

    // Mimic user clicking on the buttotn
    jest.runAllTimers();

    await waitFor(() => {
      expect(component.onRemoteCardClicked).toBeCalledWith('googledrive');
    });
  });

  const renderComponent = () => {
    const onRemoteCardClicked = jest.fn();
    const component = customRender(
      <RemotesListSection onRemoteCardClicked={onRemoteCardClicked} />
    );

    component.onRemoteCardClicked = onRemoteCardClicked;
    return component;
  };
});
