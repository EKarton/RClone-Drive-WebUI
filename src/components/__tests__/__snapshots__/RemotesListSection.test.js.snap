// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemotesListSection should match snapshot when api call is loading 1`] = `
<body>
  <div />
</body>
`;

exports[`RemotesListSection should match snapshot when api call was successful 1`] = `
<body>
  <div>
    <div
      class="MuiTypography-root MuiTypography-h7 remotes-list-section__header-text css-jab4h0-MuiTypography-root"
    >
      Your Remotes
    </div>
    <div>
      googledrive,onedrive,icloud
    </div>
  </div>
</body>
`;
